doc/design/programming_solver.md

```diff
@@ -0,0 +1,141 @@
+# Mathematical Programming Solver with SQLFlow
+
+Mathematical programming (aka. [mathematical optimization](https://en.wikipedia.org/wiki/Mathematical_optimization)) is the selection of a best element (with regard to some criterion) from some set of available alternatives. Solving optimization problems is widely used in fields like ecnomics and finance, computer science, engineering and researching. In this design, we try to make SQLFlow be able to solve below branches of programming problems using SQL statements:
+
+- [Linear programming](https://en.wikipedia.org/wiki/Linear_programming)
```

Comment:
Why are we targetting these types of programming challenges? Is that because pyomo, which is mentioned below, can solve these challenges?

Reply:
Yes, we are going to generate code using `pyomo` or the ant internal project, they both support these types of programming categories. For `pyomo`, we can check the supported "solvers" by: https://pyomo.readthedocs.io/en/stable/solving_pyomo_models.html#supported-solvers

Reply:
I see. Please weave the reasons into the text.  BTW, I am not sure if it is sufficient to consider only these two implementations of mathematical programming. It would be reasonable to justify like

> In this design, we consider enhancing SQLFlow to solve the following mathematical programming challenges.  We start from them because they are widely-used and there are reliable open-source implementations we can reuse.